test/core/libp2p/peer/address_repository/inmem_address_repository_test.cpp

```diff
@@ -0,0 +1,179 @@
+};
+
+TEST_F(InmemAddressRepository_Test, GarbageCollection) {
+  // @given address repository that has 2 peers, and some addresses
+  EXPECT_OUTCOME_TRUE_1(
+      db->addAddresses(p1, std::vector<Multiaddress>{ma1, ma2}, 100ms));
+  EXPECT_OUTCOME_TRUE_1(
+      db->addAddresses(p1, std::vector<Multiaddress>{ma3, ma4}, 400ms));
+  EXPECT_OUTCOME_TRUE_1(
+      db->upsertAddresses(p2, std::vector<Multiaddress>{ma4}, 100ms));
+
+  // @when no collectGarbage is called
+  {
+    EXPECT_OUTCOME_TRUE_2(v1, db->getAddresses(p1));
+    EXPECT_OUTCOME_TRUE_2(v2, db->getAddresses(p2));
+
+    // @then initial state is initial
+    EXPECT_EQ(v1.size(), 4);
+    EXPECT_EQ(v2.size(), 1);
+  }
+
+  // @when first collect garbage is called
+  collectGarbage();
+
+  {
+    EXPECT_OUTCOME_TRUE_2(v1, db->getAddresses(p1));
+    EXPECT_OUTCOME_TRUE_2(v2, db->getAddresses(p2));
+
+    // @then no addresses are evicted
+    EXPECT_EQ(v1.size(), 4);
+    EXPECT_EQ(v2.size(), 1);
+  }
+
+  // @when second collect garbage is called in 200ms
+  std::this_thread::sleep_for(200ms);
+  collectGarbage();
+  // ma1 and ma2 for p1 should be evicted by now
+
+  {
+    // @then p1 has evicted 2 addresses
+    EXPECT_OUTCOME_TRUE_2(v1, db->getAddresses(p1));
+    EXPECT_EQ(v1.size(), 2);
+
+    // @and p2 has been evicted completely
+    auto v2 = db->getAddresses(p2);
+    EXPECT_FALSE(v2);
+    // peers without addresses are removed... so we can't find this peer
+    EXPECT_EQ(v2.error().value(), (int)PeerError::NotFound);
+  }
+
+  // @when clear p1 addresses
+  db->clear(p1);
+
+  {
+    // @then p1 is not evicted, but all its addresses are
+    // since we intentionally cleared addresses of this peer, we do not evict
+    // this peer from the list of known peers up to the next garbage collection
+    EXPECT_OUTCOME_TRUE_2(v1, db->getAddresses(p1));
+    EXPECT_EQ(v1.size(), 0);
+
+    // @and p2 is still evicted
+    auto v2 = db->getAddresses(p2);
+    EXPECT_FALSE(v2);
+    EXPECT_EQ(v2.error().value(), (int)PeerError::NotFound);
+  }
+
+  // @when third collect garbage is called
+  collectGarbage();
+
+  {
+    // @then both p1 and p2 have been evicted completely
+    // last garbage collection removed all peers that do not have addresses
+    for (const auto &it : {p1, p2}) {
+      auto v = db->getAddresses(it);
+      EXPECT_FALSE(v);
+      EXPECT_EQ(v.error().value(), (int)PeerError::NotFound);
+    }
+  }
+}
+
+/**
+ * @given Peer p1 has address m1 with ttl 100ms
+ * @when update ttl with 1000ms, then execute collectGarbage
+ * @then ttl is updated, ma1 is not evicted
+ */
+TEST_F(InmemAddressRepository_Test, UpdateAddress) {
+  EXPECT_OUTCOME_TRUE_1(
+      db->addAddresses(p1, std::vector<Multiaddress>{ma1}, 100ms));
+  EXPECT_OUTCOME_TRUE_1(
+      db->upsertAddresses(p1, std::vector<Multiaddress>{ma1}, 1000ms));
+
+  {
+    EXPECT_OUTCOME_TRUE_2(v, db->getAddresses(p1));
+    EXPECT_EQ(v.size(), 1);
+  }
+
+  std::this_thread::sleep_for(200ms);
```

Comment:
Might it be better to slightly change the repository interface to be able to mock its clock? Because as I understand, the current implementation of tests is quite slow, as it actually sleeps for some time, and slow tests is... er... pretty bad.

Reply:
200 ms is not "slow" :)

Reply:
600 ms, as I understand. There are three calls in total. Okay, here it makes sense, because it's just 600 ms and it's not simple to mock the clock, but I still do strongly dislike this practice in tests.

Reply:
Decreased ttimeouts